src/System.Management.Automation/engine/CommandCompletion/CommandCompletion.cs

```diff
@@ -537,43 +537,23 @@ private static CommandCompletion CompleteInputImpl(Ast ast, Token[] tokens, IScr
 
                 if (results == null || results.Count == 0)
                 {
-                    /* BROKEN code commented out, fix sometime
                     // If we were invoked from TabExpansion2, we want to "remove" TabExpansion2 and anything it calls
                     // from our results.  We do this by faking out the session so that TabExpansion2 isn't anywhere to be found.
-                    MutableTuple tupleForFrameToSkipPast = null;
-                    foreach (var stackEntry in context.Debugger.GetCallStack())
+                    SessionStateScope scopeToRestore;
+                    if (context.CurrentCommandProcessor.Command.CommandInfo.Name.Equals("TabExpansion2", StringComparison.OrdinalIgnoreCase))
                     {
-                        dynamic stackEntryAsPSObj = PSObject.AsPSObject(stackEntry);
-                        if (stackEntryAsPSObj.Command.Equals("TabExpansion2", StringComparison.OrdinalIgnoreCase))
-                        {
-                            tupleForFrameToSkipPast = stackEntry.FunctionContext._localsTuple;
-                            break;
-                        }
+                        scopeToRestore = context.EngineSessionState.CurrentScope;
+                        context.EngineSessionState.CurrentScope = scopeToRestore.Parent;
```

Comment:
If `TabExpansion2` is defined in a module this will revert to the module's `script` scope rather than the actual previous scope.

Also if `TabExpansion2` is dot sourced from global, `Parent` will be null here

Reply:
> If `TabExpansion2` is defined in a module then this will revert to the module's `script` scope rather than the actual previous scope.

What's the difference? If I look at the session state API it seems to simply use the Parent property to move to different scopes:

https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/engine/SessionStateScopeAPIs.cs#L128
https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/engine/SessionStateScopeEnumerator.cs#L30
https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/engine/SessionStateVariableAPIs.cs#L1791

Can you show a simple example where it would cause a problem?

> Also if `TabExpansion2` is dot sourced from global, `Parent` will be null here

How do you dot source from global? Do you mean like this: `. $Function:TabExpansion2 "$"` because if so it won't use the parent because the command name isn't "TabExpansion2".

Reply:
> What's the difference? If I look at the session state API it seems to simply use the Parent property to move to different scopes:

A `SessionState` is sort of like a `Stack` of scopes. Every module has its own stack. To move to a different stack, `EngineSessionState` needs to be assigned. Though, afaik the previous `SessionState` is only stored in a local variable inside a few different`try/finally`s.

> How do you dot source from global? Do you mean like this: `. $Function:TabExpansion2 "$"` because if so it won't use the parent because the command name isn't "TabExpansion2".

You'd do `. TabExpansion2 'etc'`, or the more likely route would be calling `PowerShell.AddCommand("TabExpansion2", useLocalScope: false)` from a host application.

Reply:
All right, thanks for the explanation. I've added a null check for the dot sourcing scenario.  
As for the module scope, would it be any different from the current behavior? If you are tab completing a variable today it's just calling `Get-Item variable:\*` from the tabexpansion2 scope which as far as I can tell retrieves variables from that scope and traverses through the `Parent` property on each scope. So with my change the only difference would be that it skips the initial TabExpansion2 scope, right?

Reply:
> Ideally we'd be able to detect that TabExpansion2 was dot sourced and abort the attempt to revert scope. I don't know of a reliable way to tell that from here though

I think you do without realizing. There's a `UseLocalScope` property in the `CurrentCommandProcessor` and you said that `PowerShell.AddCommand("TabExpansion2", useLocalScope: false)` is equivalent to dot sourcing, so can't we just use that? It seems to toggle back and forth like I would expect when I dotsource or run it normally.

Alternatively, instead of changing the scope we could just add the parent scope as an additional property to the CompletionContext object and update the variable/member completion code to use the parent scope if it's available. I don't think any other completion paths is affected by the session state but I could be wrong.

Reply:
> I think you do without realizing. There's a `UseLocalScope` property in the `CurrentCommandProcessor` and you said that `PowerShell.AddCommand("TabExpansion2", useLocalScope: false)` is equivalent to dot sourcing, so can't we just use that?

Oh yay I was wrong! 😁 I must have been thinking of when you only have the call stack, good catch!